File: src/crypto.rs
//! Cryptographic operations using ChaCha20-Poly1305.

use chacha20poly1305::{
    aead::{Aead, KeyInit},
    ChaCha20Poly1305, Nonce,
};
use rand::RngCore;
use zeroize::Zeroize;

use crate::error::{Dice52Error, Result};

/// Securely zero a byte slice
pub fn zero_bytes(bytes: &mut [u8]) {
    bytes.zeroize();
}

/// Generates n random bytes
pub fn rand_bytes(n: usize) -> Vec<u8> {
    let mut bytes = vec![0u8; n];
    rand::thread_rng().fill_bytes(&mut bytes);
    bytes
}

/// Creates a nonce including session ID, epoch and message number
fn make_nonce(sid: u32, epoch: u64, ctr: u64) -> ([u8; 12], Nonce) {
    let mut n = [0u8; 12];
    n[0..4].copy_from_slice(&sid.to_be_bytes());
    n[4..8].copy_from_slice(&(epoch as u32).to_be_bytes());
    n[8..12].copy_from_slice(&(ctr as u32).to_be_bytes());
    let nonce = Nonce::from(n);
    (n, nonce)
}

/// Encrypts plaintext using ChaCha20-Poly1305 (Section 11.2)
///
/// # Arguments
/// * `mk` - Message key (32 bytes)
/// * `sid` - Session ID
/// * `epoch` - Current epoch
/// * `ctr` - Message counter
/// * `ad` - Associated data
/// * `pt` - Plaintext
///
/// # Returns
/// Ciphertext with authentication tag
///
/// Note: The caller is responsible for zeroing the message key after use.
pub fn encrypt(mk: &[u8], sid: u32, epoch: u64, ctr: u64, ad: &[u8], pt: &[u8]) -> Vec<u8> {
    let cipher = ChaCha20Poly1305::new_from_slice(mk).expect("Invalid key length");
    let (mut n, nonce) = make_nonce(sid, epoch, ctr);

    let ct = cipher
        .encrypt(&nonce, chacha20poly1305::aead::Payload { msg: pt, aad: ad })
        .expect("Encryption should not fail");

    n.zeroize(); // Zero nonce after use
    ct
}

/// Decrypts ciphertext using ChaCha20-Poly1305 (Section 12)
///
/// # Arguments
/// * `mk` - Message key (32 bytes)
/// * `sid` - Session ID
/// * `epoch` - Current epoch
/// * `ctr` - Message counter
/// * `ad` - Associated data
/// * `ct` - Ciphertext with authentication tag
///
/// # Returns
/// Decrypted plaintext or error
///
/// Note: The caller is responsible for zeroing the message key after use.
pub fn decrypt(mk: &[u8], sid: u32, epoch: u64, ctr: u64, ad: &[u8], ct: &[u8]) -> Result<Vec<u8>> {
    let cipher = ChaCha20Poly1305::new_from_slice(mk).expect("Invalid key length");
    let (mut n, nonce) = make_nonce(sid, epoch, ctr);

    let result = cipher
        .decrypt(&nonce, chacha20poly1305::aead::Payload { msg: ct, aad: ad })
        .map_err(|e| Dice52Error::DecryptionFailed(e.to_string()));

    n.zeroize(); // Zero nonce after use
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encrypt_decrypt_roundtrip() {
        let mk = rand_bytes(32);
        let sid = 1u32;
        let epoch = 0u64;
        let ctr = 0u64;
        let ad = b"test associated data";
        let pt = b"Hello, quantum world!";

        let ct = encrypt(&mk, sid, epoch, ctr, ad, pt);
        let decrypted = decrypt(&mk, sid, epoch, ctr, ad, &ct).unwrap();

        assert_eq!(decrypted, pt);
    }

    #[test]
    fn test_decrypt_with_wrong_key_fails() {
        let mk = rand_bytes(32);
        let wrong_mk = rand_bytes(32);
        let sid = 1u32;
        let epoch = 0u64;
        let ctr = 0u64;
        let ad = b"test";
        let pt = b"secret";

        let ct = encrypt(&mk, sid, epoch, ctr, ad, pt);
        let result = decrypt(&wrong_mk, sid, epoch, ctr, ad, &ct);

        assert!(result.is_err());
    }
}

File: clients/java/src/main/java/io/dice52/Crypto.java
package io.dice52;

import org.bouncycastle.crypto.engines.ChaCha7539Engine;
import org.bouncycastle.crypto.macs.Poly1305;
import org.bouncycastle.crypto.modes.ChaCha20Poly1305;
import org.bouncycastle.crypto.params.AEADParameters;
import org.bouncycastle.crypto.params.KeyParameter;

import java.nio.ByteBuffer;
import java.security.SecureRandom;
import java.util.Arrays;

/**
 * Cryptographic operations using ChaCha20-Poly1305.
 */
public class Crypto {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private Crypto() {
        // Utility class
    }

    /**
     * Generate n random bytes using secure random.
     */
    public static byte[] randBytes(int n) {
        byte[] bytes = new byte[n];
        SECURE_RANDOM.nextBytes(bytes);
        return bytes;
    }

    /**
     * Securely zero a byte array.
     */
    public static void zeroBytes(byte[] bytes) {
        if (bytes != null) {
            Arrays.fill(bytes, (byte) 0);
        }
    }

    /**
     * Create a nonce including session ID, epoch and message number.
     * Nonce is 96 bits (12 bytes).
     */
    public static byte[] makeNonce(int sid, long epoch, long ctr) {
        ByteBuffer buffer = ByteBuffer.allocate(12);
        buffer.putInt(sid);
        buffer.putInt((int) epoch);
        buffer.putInt((int) ctr);
        return buffer.array();
    }

    /**
     * Encrypt plaintext using ChaCha20-Poly1305 (Section 11.2).
     */
    public static byte[] encrypt(byte[] mk, int sid, long epoch, long ctr, byte[] ad, byte[] pt) {
        try {
            byte[] nonce = makeNonce(sid, epoch, ctr);

            ChaCha20Poly1305 cipher = new ChaCha20Poly1305();
            AEADParameters params = new AEADParameters(
                    new KeyParameter(mk),
                    128, // Tag size in bits
                    nonce,
                    ad);
            cipher.init(true, params);

            byte[] output = new byte[cipher.getOutputSize(pt.length)];
            int len = cipher.processBytes(pt, 0, pt.length, output, 0);
            cipher.doFinal(output, len);

            zeroBytes(nonce);
            return output;
        } catch (Exception e) {
            throw new RuntimeException("Encryption failed", e);
        }
    }

    /**
     * Decrypt ciphertext using ChaCha20-Poly1305 (Section 12).
     */
    public static byte[] decrypt(byte[] mk, int sid, long epoch, long ctr, byte[] ad, byte[] ct)
            throws Dice52Exception.DecryptionFailed {
        try {
            if (ct.length < 16) {
                throw new Dice52Exception.DecryptionFailed("Ciphertext too short");
            }

            byte[] nonce = makeNonce(sid, epoch, ctr);

            ChaCha20Poly1305 cipher = new ChaCha20Poly1305();
            AEADParameters params = new AEADParameters(
                    new KeyParameter(mk),
                    128, // Tag size in bits
                    nonce,
                    ad);
            cipher.init(false, params);

            byte[] output = new byte[cipher.getOutputSize(ct.length)];
            int len = cipher.processBytes(ct, 0, ct.length, output, 0);
            cipher.doFinal(output, len);

            zeroBytes(nonce);
            return output;
        } catch (Dice52Exception.DecryptionFailed e) {
            throw e;
        } catch (Exception e) {
            throw new Dice52Exception.DecryptionFailed("Decryption failed: " + e.getMessage(), e);
        }
    }
}
